addons/web_editor/controllers/main.py

```diff
@@ -521,64 +523,97 @@ def modify_image(self, attachment, res_model=None, res_id=None, name=None, data=
             '1': '#3AADAA',
             '2': '#7C6576',
             '3': '#F6F6F6',
             '4': '#FFFFFF',
             '5': '#383E45',
         }
+
+        for key, value in options.items():
+            colorMatch = re.match('^c([1-5])$', key)
+            if colorMatch:
+                # Check that color is hex or rgb(a) to prevent arbitrary injection
+                if not re.match(r'(?i)^#[0-9A-F]{6,8}$|^rgba?\(\d{1,3},\d{1,3},\d{1,3}(?:,[0-9.]{1,4})?\)$', value.replace(' ', '')):
+                    raise werkzeug.exceptions.BadRequest()
+                user_colors.append([tools.html_escape(value), colorMatch.group(1)])
+            else:
+                svg_options[key] = value
+
         color_mapping = {default_palette[palette_number]: color for color, palette_number in user_colors}
         # create a case-insensitive regex to match all the colors to replace, eg: '(?i)(#3AADAA)|(#7C6576)'
         regex = '(?i)%s' % '|'.join('(%s)' % color for color in color_mapping.keys())
 
         def subber(match):
             key = match.group().upper()
             return color_mapping[key] if key in color_mapping else key
-        svg = re.sub(regex, subber, svg)
+        return re.sub(regex, subber, svg), svg_options
+
+    @http.route(['/web_editor/shape/<module>/<path:filename>'], type='http', auth="public", website=True)
+    def shape(self, module, filename, **kwargs):
+        """
+        Returns a color-customized svg (background shape or illustration).
+        """
+        svg = None
+        if module == 'illustration':
+            attachment = request.env['ir.attachment'].sudo().browse(unslug(filename)[1])
+            if (not attachment.exists()
+                    or attachment.type != 'binary'
+                    or not attachment.public
+                    or not attachment.url.startswith(request.httprequest.path)):
+                raise werkzeug.exceptions.NotFound()
+            svg = b64decode(attachment.datas).decode('utf-8')
+        else:
+            svg = self._get_shape_svg(module, 'static', 'shapes', filename)
+
+        svg, options = self._update_svg_colors(kwargs, svg)
+        flip_value = options.get('flip', False)
+        if flip_value == 'x':
+            svg = svg.replace('<svg ', '<svg style="transform: scaleX(-1);" ')
+        elif flip_value == 'y':
+            svg = svg.replace('<svg ', '<svg style="transform: scaleY(-1)" ')
+        elif flip_value == 'xy':
+            svg = svg.replace('<svg ', '<svg style="transform: scale(-1)" ')
 
         return request.make_response(svg, [
             ('Content-type', 'image/svg+xml'),
             ('Cache-control', 'max-age=%s' % http.STATIC_CACHE_LONG),
         ])
 
+    @http.route(['/web_editor/image_shape/<module>/<path:filename>'], type='http', auth="public", website=True)
+    def image_shape(self, module, filename, img_key, **kwargs):
+        svg = self._get_shape_svg(module, 'static', 'image_shapes', filename)
+        if img_key:
```

Comment:
Kinda strange to have a required attribute as additional parameter, maybe change the URL form to:

`/web_editor/image_shape/<the image key here>/<module>/<path:filename>`

Reply:
@xO-Tx The point of the URL change, was to be able to remove that condition, you now always have an `img_key`.